File: factory_method/go/app/usecases/factory.go
package usecases

import (
	"factory_method/app/service/agoda"
	"factory_method/app/service/booking"
	"factory_method/app/service/expedia"
	"factory_method/domain"
)

func NewReservation(rsvnType domain.ReservationType) domain.Reservation {
	switch rsvnType {
	case domain.Agoda:
		return agoda.NewAgodaReservation()
	case domain.Expedia:
		return expedia.NewExpediaReservation()
	case domain.Booking:
		return booking.NewBookingReservation()
	}

	return nil
}

File: factory_method/kotlin/src/main/kotlin/app/usecases/Factory.kt
package app.usecases

import app.service.agoda.Agoda
import app.service.booking.Booking
import app.service.expedia.Expedia
import domain.Reservation
import domain.ReservationType

class Factory {
    companion object {
        fun GetReservationWay(rsvnType: ReservationType): Reservation {
            when (rsvnType) {
                ReservationType.Agoda -> return Agoda()
                ReservationType.Booking -> return Booking()
                ReservationType.Expedia -> return Expedia()
                else -> throw java.lang.RuntimeException("Error!")
            }
        }
    }
}